### PR TITLE
feat(offense): PR-124 — runner binary-binding + named egress audit + infra-failure budget

### LINKED ISSUE #124: PR5b: bind offensive-runner executable to the tool registry + named egress in audits + infra-failure budget counter
## PR5a offensive-runner — 3 deferred refinements for PR5b

Three round-3 review findings on `mcp/lib/offensive-runner.js` (#123) were deferred with reasoning. All three are on a path **no agent can reach in PR5a** (no MCP tool surface; `classify` defaults to `null`, so the runner never runs in production), and all three are PR5b's defining work — PR5b wires the live tool + the tool registry + named egress on top.

### 1. Bind `toolArgv[0]` (the container binary) to the tool registry
PR5a shape-guards the executable to a bare `[a-zA-Z0-9._-]` token (no path/metacharacters), but a live caller forwarding agent-influenced argv could still choose another bare binary present in the digest-pinned image (`sh`, `curl`) and have it run + audited as the supplied `toolId`. The real fix is to validate `toolArgv[0]` against the **server-side tool spec's declared binary for `toolId`** — which requires PR5b's tool registry (`offensive-tools.js` / the `bob_offensive_*` factory).

### 2. Record the selected egress profile identity in the audit
PR5a records the audit row's egress as the literal `"proxy"` when a proxy url is supplied. PR5b resolves session-bound **named** egress profiles (like the existing producers via `resolveAndAssertSessionEgressIdentity`); record the resolved profile name/identity so http-audit / circuit / report consumers can tell which egress the wide-open container used.

### 3. Don't consume the run budget on infra failure
The budget slot is reserved (counter incremented) **before** spawn, so repeated infrastructure failures (network-create / spawn errors) could consume the session's 200-run cap. The reserve-before-spawn is the atomic race fix (count + reserve under the lock); counting attempts is the fail-closed policy. PR5b: a separate infra-failure counter that does not consume the offensive-run budget, while keeping the reservation atomic for genuine runs.

### Acceptance
- `toolArgv[0]` validated against the registry's declared binary for `toolId`.
- Named egress profile identity recorded in the offensive-run audit row.
- Infra failures distinguished from real runs in the budget accounting, without reopening the count/reserve race.

> Note: the round-3 budget *trim-bypass* (counting from the 5000-trimmed `http-audit.jsonl`) and the *perf* concern (full-log read under the lock) were **fixed in #123** by switching the budget to a monotonic untrimmable counter file — not part of this deferral.

### PR DESCRIPTION
**Pull request by vmihalis**
## What

The three deferred PR5a runner refinements from **#124**, applied to the still-inert
`offensive-runner.js` with stubbed-docker tests. This is `PR-124` in the locked offensive
sequence (`PR-IMAGE` #125 ✅ → **PR-124** → `PR5b-tool`).

**The seam:** two of the three findings "require the registry" per #124. PR-124 makes the
runner **enforce** binary-binding + named-egress given its inputs; `PR5b-tool`'s tool
registry (`offensive-tools.js` / the `bob_offensive_*` factory) will **provide** those
inputs (the declared binary + the resolved egress name). The runner is a pure function of
its inputs; this PR completes + enforces its input contract.

### 1. Bind `toolArgv[0]` to the declared binary (`flagSpec.binary`) — #124-1
PR5a only shape-guarded the executable to a bare `[a-zA-Z0-9._-]` token, so a live caller
forwarding agent-influenced argv could still pick a *different* in-image binary (`sh`,
`curl`) and have it run + audited under the supplied `tool_id`. Now, when `flagSpec.binary`
is declared, `toolArgv[0]` must equal it (and the declared binary is itself shape-validated
— can't smuggle a path). When no binary is declared (the PR5a inert path) the bare-token
check is the only constraint, so this is back-compatible. PR5b's registry sets
`flagSpec.binary` per `tool_id` from server-side config (never agent input).

### 2. Named egress in the `CONTAINER_RUN` audit — #124-2
PR5a recorded the audit row's egress as the literal `"proxy"`. Now the runner takes
`egressProfileName` (the resolved profile identity) and stamps **that** into the audit
row's `egress_profile`, so circuit-breaker / report consumers can tell which egress the
wide-open container used. A proxied run **must** name its profile — an anonymous proxy is
rejected. PR5b resolves session-bound named profiles via
`resolveAndAssertSessionEgressIdentity` and passes the name here.

### 3. Infra failures don't consume the run budget — #124-3
The budget slot is reserved (counter incremented) **before** spawn — the #123 atomic
TOCTOU fix. So repeated infrastructure failures (network-create / spawn errors, where the
container never started) could burn the 200-run cap. Now those two outcomes **refund** the
reserved slot and bump a **separate, capped** infra-failure counter.

**Why reserve-then-refund (not "don't count"):** you must reserve *before* spawn to keep
the reservation atomic (two concurrent agents reading a stale count would otherwise both
pass the gate), and you can't know an attempt is an infra failure until *after* the
network/spawn step. So refunding post-hoc under the lock is the only way to "not consume
the budget on infra failure" **while keeping the reservation atomic** — exactly what #124
asks for. A **timeout is NOT refunded**: the container actually ran (consumed CPU/net + a
real slot) and only failed to finish — a genuine run. The separate cap
(`MAX_OFFENSIVE_INFRA_FAILURES = 50`) preserves the total-attempts bound the reserve used
to provide (so a daemon/target that fails every spawn can't spin unbounded docker attempts).

## Scope (unchanged from PR5a)
Still **no MCP tool surface, no `classify`, no registry** — the runner remains inert in
production (`classify` defaults to `null`, never signs). No generated surface changes (the
runner isn't a `TOOLS` entry). The registry + live tool + `command_hash` binding + sound
classify are `PR5b-tool`.

## Tests
`test/offensive-runner.test.js`: **24 → 33** (+9). New coverage: declared-binary match /
mismatch / malformed (pure + runner), egress-name required / malformed / recorded-in-audit
/ no-literal-"proxy", infra-failure refund (network-create + spawn) + separate counter,
timeout-consumes-budget, infra-cap-exhausted blocks.

- `npm run test:mcp` — **2696 / 0**
- `npm run check:syntax` — clean
- `npm run test:prompts` — **181 / 0**

Closes #124.

## Summary by CodeRabbit

* **New Features**
  * Added egress-profile support for proxy-based runs, including validation and audit logging that records the resolved profile name.
* **Bug Fixes**
  * Improved run-budget accounting: infrastructure failures are tracked and capped separately, refunded from the genuine run budget, and enforced to prevent new runs once the infra cap is reached.
  * Tightened allowlist enforcement by requiring an exact match for declared binary names.
* **Tests**
  * Expanded coverage for egress-profile/audit behavior, proxy validation rules, and fail-closed infra counter handling and edge-case scenarios.

### CODE REVIEW

**Comment by vmihalis** [comment]
@codex review

@coderabbitai review

File: mcp/lib/offensive-runner.js [hunk 1]
```diff
@@ -406,7 +474,10 @@ async function runOffensiveTool({
   const auditUrl = primaryUrl ? String(primaryUrl.value) : `https://${domain}/`;
   const reservation = withSessionLock(domain, () => {
     if (offensiveRunCount(domain) >= MAX_OFFENSIVE_RUNS) return { ok: false, reason: "offensive_run_budget_exhausted" };
-    const auditOk = auditConfirmRequest({ domain, surfaceId: null, method: OFFENSIVE_RUN_AUDIT_METHOD, url: auditUrl, egressProfile: egressProxyUrl ? "proxy" : null, status: null, scopeDecision: "allowed", error: null, startedAt, toolId });
+    // #124-3: infra failures are refunded from the run budget, so cap them separately to
+    // keep the total-attempts bound the reserve-before-spawn used to provide.
+    if (offensiveInfraFailureCount(domain) >= MAX_OFFENSIVE_INFRA_FAILURES) return { ok: false, reason: "offensive_infra_failure_budget_exhausted" };
```
[the remark this answers is not in the document]
Comment (vmihalis) [reply mcp/lib/offensive-runner.js:537]: Fixed in 17f8e7a. The infra slot is now reserved **pessimistically under the reservation lock** before docker spawns, so a concurrent burst can't pass a stale count and overshoot the cap. A genuine run releases its slot at a single point right after the `spawn_error` check (every outcome past that point is a genuine run); an infra failure keeps its slot as the recorded failure. Net per outcome is unchanged, so the existing refund/timeout tests still pass — and the cap now also bounds concurrent wide-open containers to MAX_OFFENSIVE_INFRA_FAILURES. Added a test asserting the slot is held during the run and released on a genuine outcome.

**Comment by vmihalis** [comment]
@codex review

@coderabbitai review

**Comment by vmihalis** [comment]
## Round-2 review triage (commit `4d3ef5b`)

Round 1's concurrency fix (pessimistic infra reservation) drew a cluster of round-2 findings. On review they converge on one conclusion: **the pessimistic approach traded an unreachable tightness for real fragility**, so I reverted it and hardened the genuinely cross-cutting issues.

### Fixed
- **P1 — symlinked counter leaves** (`#3433039286`): `readCounterFile`/`writeCounterFile` now use `O_NOFOLLOW` + regular-file/`nlink` discipline. A same-UID symlinked counter leaf can no longer redirect a write outside session files, nor reset the budget by pointing at a `"0"` file. Hardens **both** counters (incl. the pre-existing run budget).
- **P2 — `default` profile while proxying** (`#3433039267`): reject `egressProfileName: "default"` when a proxy URL is set (`egress-profiles.js` pins `default` to `proxy_url: null`, so it would misbind a proxied run as direct egress).
- **cap-bypass on partial write** (brutalist `#3432958539`/`#3432958545`): the refund now records the infra failure **first** and only frees the run slot if it persisted — a storage failure can never free a slot without counting the failure. Worst case is an over-count, never an under-count.
- **P2 lock-contention on release** (`#3433039275`) + **crash/no-TTL leak** (`#3433052891`) + **release/rollback return-value discards** (`#3433052896`/`#3433052904`/`#3433052910`/`#3433052895`): all were consequences of the per-genuine-run release. **Removed by the revert** — post-hoc accounting touches the lock only on a real network-create/spawn failure (rare), and that path is wrapped best-effort against the fail-fast lock-busy throw.

### Declined (with rationale)
- **Concurrency-tightness of the infra cap** (the original `#3432936055`): I'm **not** re-adding the pessimistic reservation. The concurrent burst is already bounded by the **run budget** (200) — each in-flight run holds a run slot until refund, so total docker spawns in flight ≤ `MAX_OFFENSIVE_RUNS`, the real anti-thrash bound. The infra cap bounds *sequential* spin. Enforcing exact-cap under a 50+-parallel burst requires the pessimistic hold, which your own round-2 findings show introduces a **permanent crash-leak** (no TTL) and a **per-run fail-fast-lock contention leak** — strictly worse, and the 50+-concurrency trigger is unreachable at Bob's wave concurrency (~12). Net: the simpler post-hoc design is the better trade.
- **egress name↔proxy-URL binding** (`#3432958556`/`#3433052889`): `egressProfileName` remains caller-supplied. #124-2 explicitly scopes egress **resolution** to PR5b (which calls `resolveAndAssertSessionEgressIdentity`, binding name↔url server-side); PR-124's mandate is to *record* the resolved name. The runner can't resolve the named-profile registry here, but it now rejects the one universally-known inconsistency (`default` + proxy). Full binding lands with PR5b.

`test:mcp` **2698/0**, runner suite **35/0**, `check:syntax` clean.

@codex review
@coderabbitai review

File: mcp/lib/offensive-runner.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (vmihalis) [reply mcp/lib/offensive-runner.js:537]: Update: **reverted** the pessimistic fix in 4d3ef5b. Round-2 review showed it introduced a permanent **crash-leak** (the infra slot is held across the docker await with no TTL, so a SIGKILL between reserve and release leaks it forever) and a **per-genuine-run fail-fast-lock contention leak** — both strictly worse than a bounded overshoot. The concurrent burst is already bounded by the run budget (each in-flight run holds a run slot until refund, so docker spawns in flight ≤ MAX_OFFENSIVE_RUNS); the infra cap bounds sequential spin. I'm declining the exact-cap-under-a-50+-concurrent-burst tightening: it's unreachable at Bob's wave concurrency (~12) and the pessimism is a net-negative trade. Full rationale in the PR triage comment.

**Comment by vmihalis** [comment]
## Round-3 triage (commit `5e53d3c`)

All 3 round-3 findings fixed (they were on the round-2 counter hardening + a real `--pull=never` gap):

- **P1 whole-file validation** (Codex + CodeRabbit + brutalist): `readCounterFile` now `fstat`-size-gates (>32 bytes = tampered → fail closed) and validates the **whole** file, so a padded `"0"×32 + junk` leaf can no longer parse as `0` and reset the run/infra budget.
- **P2 FIFO** (Codex): both counter opens add `O_NONBLOCK`, so a FIFO/device leaf returns immediately (no hang while holding the session lock) and the `fstat` rejects it → fail closed.
- **P2 docker exit 125/126/127** (Codex): the daemon/exec "couldn't start the container" exits (esp. a `--pull=never` image miss, likely before the operator mints the image) now **refund + count as infra failures** instead of burning the 200-run budget.
- **test** (CodeRabbit): the symlink test now points at a valid `"0"` so it proves `O_NOFOLLOW` (not the target's value) is what causes the fail-closed.

### Declined (logged)
- **egressProfileName-without-proxy** (brutalist): the runner can't resolve the named-profile registry in PR-124 (PR5b's `resolveAndAssertSessionEgressIdentity` binds name↔url server-side); it enforces the one knowable invariant — `default` ⟺ no-proxy. Same documented seam as the earlier egress decline.

`test:mcp` **2700/0**, runner suite **37/0**, `check:syntax` clean.

@codex review
@coderabbitai review

**Comment by vmihalis** [comment]
## Round-4 triage (commit `5e53d3c`) — declines, ready to merge

Round 4 is clean on the primary gates: **CodeRabbit passed with no findings**, and Codex posted no new inline findings (all prior P1/P2s resolved). The remaining 3 items are brutalist-CI mediums, all defensibly declined:

### Declined (logged)
1. **egress name↔proxy-URL binding** (brutalist Claude + Codex, two angles): the runner records the caller-supplied `egressProfileName` without resolving it against the proxy URL. This is the **documented PR5b seam** — #124-2 explicitly scopes egress *resolution* to PR5b (`resolveAndAssertSessionEgressIdentity`, which binds name↔url + sets `proxy_configured`/identity-hash server-side). PR-124's mandate is to record the resolved name + enforce the one invariant the runner can know without the registry (`default` ⟺ no-proxy, now rejected). Pulling full resolution into PR-124 is the exact PR5b scope-creep the "sequence, don't bundle" decision avoided.

2. **CONTAINER_RUN audit row not corrected on infra failure** (brutalist Codex, correctness): that row is a **pre-run, fail-closed breaker-visibility gate** (#123 — a run that can't be recorded for the breaker is not allowed to proceed). `scope_decision: "allowed"` is the *scope* verdict (target is in-scope — accurate), not a success claim; the run's outcome (infra failure) is recorded in the infra-failure counter + returned masked to the agent. A local docker/infra failure is correctly **not** a target-side circuit event (not a 403/429/scope-block/timeout), so the uncorrected row does not falsely trip the breaker — and appending a correction row would double-count attempts in the breaker's view. Outcome-accurate audit rows are a reasonable enhancement once a report consumer exists (PR5b+); the runner is inert in PR-124.

All CI green (`test (20)`/`test (22)`/`brutalist`/`Bob Diff Review`/CodeRabbit). `test:mcp` **2700/0**, runner suite **37/0**.

Merging on the basis of: clean primary review (CodeRabbit + Codex inline) + the above logged declines for the supplementary critic.
